tests/system/utils.py

```diff
@@ -39,3 +40,121 @@ def get_resource(resource):
                 raise Exception
         except Exception:
             raise DCOSException("Can't read from resource: {0}. Please check that it exists.".format(resource))
+
+
+class FaultDomain:
+    """
+    High-level representation of a fault domain
```

Comment:
What does that mean? What do you think about using a `namedtuple` instead?

Reply:
I am indifferent on that. However I personally like a fully described type definition more than using generators like `namedtuple` .. but that might be more my C++ background 😄

Reply:
Oh, now I remember why. The fault domain is obtained by doing `master.state().get('domain')` on the mesos API. And the response is quite irregular (ex. it can be `None` if there is no fault domain, or it could be an empty object etc.)

That's why I created the `FaultDomain` class, as a wrapper to the response. So, by doing:
```
FaultDomain(master.state().get('domain'))
```
you always get a well-known reference to the fault domain

Reply:
Reading the code a little more I see how we can improve this. The `FaultDomain` has a region and zones. That should ease your use cases a lot.

Reply:
I didn't get the last part. Can you elaborate on _how_ it can ease the use cases?

Reply:
```python
class FaultDomain:

    def __init__(self, ...):
        self.region = '...'
        self.zones = list(...)

def get_biggest_cluster_region():
    max(get_all_cluster_domains(), key=lambda d: len(d.zones))

@common.require(len(get_biggest_cluster_region().zones) > 1, "Cluster must have at least 2 zones in the biggest region")
def test_faultdomains_zone_unique():
    """Tests if the applications with a region fixed and a ["@zone", "UNIQUE"] constraint correctly starts
       the apps across all zones in the given region
    """
    # Find out the biggest region in the cluster
    domain = get_biggest_cluster_region()
    zones = domain.zones
    ...
```

Maybe I'm a little confused about the relation between domain, region and zone. When you say `get_biggest_cluster_region()` don't you mean a domain which has the most zones?

Reply:
Inclined to believe that this is to find the remote zone for testing.

Couldn't we just find and use any region that:

- has the requisite number of zones
- isn't the default region

?

Reply:
So, I still don't get the full idea, but let me elaborate:
1. The cluster is separated in `regions` 
2. Each region is further separated in `zones`
3. Each agent is launched in a `zone`
4. The "domain" of an agent is the tuple `(region, domain)` where it's launched.
5. This domain "tuple" is represented with the `FaultDomain` class. (that's why I don't get your idea of using `zones = list(...)`)

![regions-zones](https://user-images.githubusercontent.com/883486/46212726-561e9180-c33f-11e8-845e-b17909bac07c.png)